Post-docs-drive edits to Application Architecture (#295)
holochain/docs-pages: src/concepts/2_application_architecture.md

When a client calls a function in a hApp, it specifies the **cell ID**, which is the hash of the DNA plus the agent ID, along with the zome name, function name, and function input payload.

### DNA

A bundle of executable code that makes a unit of functionality in a hApp is called a **DNA**. You can think of it like a [microservice](https://en.wikipedia.org/wiki/Microservices) that creates a data access and integrity layer around personal and shared data. It serves as the ‘rules of the game’ against which peers can do validation and enforcement.

The DNA can also contain metadata: a name, description, unique ID, and **properties**. The unique ID and properties can be changed either in a text editor or at installation time. The unique ID, on the other hand, can be changed to **clone** a DNA, creating a new cell with identical functionality but an entirely separate history, network, and shared database. The properties, on the other hand, can also be changed to clone a DNA, but also direct the executable code to change the new cell's runtime behavior (similar to configuration parameters). 

> In fact, even the slightest alteration of any part of the DNA will do this. Consider source code changes carefully, because each modification will create a new DNA with a new cells interacting in a separate network. This may require some sort of migration strategy to move or access data between old and new cells, which means that you should introduce changes carefully.

Author: @Connoropolous This is my latest iteration of the commentary on migration. What do you think? Leaving implementation details unspecified/open-ended (I sorta think special migration zome functions + bridging are the answer, but I suspect changing a DNA means you need to migrate to a totally new hApp and I don't know if one hApp can call another's functions).

Reviewer: Good, but there's repetition of "introduce changes carefully", maybe once more than necessary

Author: oh! good catch

### Zome

![](../../img/concepts/2.5-zome.png)

The executable code modules in a DNA are called **zomes** (short for chromosomes), each with its own name like `profile` or `chat`. The zomes define the core business logic in a DNA, exposing their functions to the conductor. Some of these functions are 'hooks' called automatically by Holochain, such as an initialization function or validation functions related to data types defined in the zome.

> Other functions are invented by the developer, have arbitrary names, and define the zome’s public API. The conductor [makes this API available](../8_calls_capabilities/) to other zomes within the DNA, other DNAs within the hApp, and, as mentioned earlier, clients running on the participant's machine and other agents on the DNA's network. The developer can give a participant the ability to control access to their cell's API via [capabilities](../8_calls_capabilities/).

Author: @Connoropolous do you think it's long-winded to introduce all these details here? I'm wondering if this should be saved for the Guidebook -- at the same time, devs need to know how their code can be called and probably want to know how access can be controleld too.

Reviewer: looks good

In some ways, you can think of the conductor as a web application server, but one that runs on every user's device. It is called the conductor because in one sense it ['leads the orchestra'](https://en.wikipedia.org/wiki/Conducting), and in another sense because it has good ['conductivity'](https://en.wikipedia.org/wiki/Electrical_conductor).

hApps communicate with each other privately and securely in peer-to-peer networks thanks to the conductor. The conductor can manage more than one set of private/public key pairs, and the same private/public key pair can be used with various hApps.

### hApp

![](../../img/concepts/2.10-network.png)

A Holochain application or **hApp** allows a single user or **agent** to easily install and manage a suite of functionality, such as chat, accounting, or project management. A hApp is often made of multiple DNAs, each providing an aspect of functionality, and it has a **slot** for each.

Some slots can be filled with multiple cells **cloned** from the same DNA. This is useful for creating separate spaces inside a single hApp --- individual project workspaces or private chat channels, for example.

### Cell

> **Cells** occupy slots in a hApp. Each cell is a combination of a user's unique cryptographic key (their **agent ID**) and a running DNA. Within one user's instance of a hApp, all cells share the same agent ID.  Each cell acts as the user’s personal agent --- every piece of data that it creates or message it sends, it does so from the perspective of that agent.

Reviewer: " every piece of data that it creates or message it sends, it does so from the perspective of that agent." this could be a link to 'source chain' in the glossary or something. I would say introduce the term, but omg there's so many terms to throw around :p

Author: I know; soo soo many terms. Grappling with how to progressively introduce ideas without causing the reader to go madly off in all directions.

Each cell stores its own agent history and is a member of a separate, isolated network within the hApp with its own data store.

> When a client calls a function in a hApp, it specifies the **cell ID**, which is the hash of the DNA plus the agent ID, along with the zome name, function name, and of course the input payload.

Reviewer: ```suggestion
When a client makes a call to the API of a hApp, it has to specify the **cell ID**, which is the hash of the DNA plus the agent ID.
```

those other inputs can be left and introduced level by level?

Author: I also wonder if this is an implementation detail and should be pulled out entirely -- trying to be light on details and heavy on concepts. However... there is no guidebook yet, so the core concepts have to suffice. Probs what I'll do is move it to the very end, once all the bits have been introduced in detail.

Reviewer: I was thinking that at the end of this you can just link off to the (now even more flushed out) happ-client-call-tutorial which would give you details.

### DNA

A bundle of executable code that makes a unit of functionality in a hApp is called a **DNA**. It serves as the ‘rules of the game’ against which peers can do peer-to-peer validation and 'enforcement'. You can think of it like a [microservice](https://en.wikipedia.org/wiki/Microservices).

The DNA is made of one or more executable code submodules called zomes, each with its own name like `profile` or `chat`. It also contains metadata such as a name and description, along with an arbitrary unique ID and **properties** (constants that the executable code can access in order to change its behavior).

> The unique ID and properties can be changed either in a text editor, at installation time, or at cloning time. Due to the properties of cryptographic hashing, even the slightest alteration of either the code, properties, or unique ID will change the hash, forming a cell with an entirely new history and peer-to-peer network. This means that you should introduce changes very carefully.

Reviewer: ```suggestion
The unique ID and properties can be changed either in a text editor, at installation time, or at cloning time. Due to the properties of cryptographic hashing, even the slightest alteration of either the code, properties, or unique ID will change the hash, forming a cell with an entirely new history and peer-to-peer network. This means that you should introduce changes very carefully. Software updates may use a pattern similar, but different, to traditional database migrations.
```

Author: Could you elaborate more on this? I feel like it's an important thing to bring up, and this is the right place to talk about it, but I'm having a hard time grasping what this pattern looks like.

Reviewer: well, what we're learning with Acorn is that the handling of software updates, if they're to include DNA updates, are entirely up to us as the software developers to define, with little to no guidance from Holochain. We currently are providing end users with JSON data dump exports of their projects, that they can re-import into a new version... then the version will have to be responsible for realizing what the schema is they're coming from so that it can import it correctly.

Reviewer: Not posing this is as a criticism ^

Author: I don't want to get too far into the weeds re: migration -- I see this as a job for a future Guidebook -- but I do want to mention something as you say. I left it open-ended re: how to implement migration; the commit isn't showing up in this PR yet though 🤷

* Each user has their own copy of the client, hApp, and conductor.
* All code is executed on behalf, and from the perspective, of the individual user.
* A DNA is instantiated into a cell for each user on their device. Each cell has a separate history and belongs to a separate private network.
* Users communicate and share data directly with one another rather than through a central server or blockchain validator network.
* Holochain is opinionated about data --- it handles all storage and retrieval. (We’ll learn about why and how in the next three articles.)
* Zomes don’t maintain any in-memory state between calls; state is maintained in the history of the cell that contains the zome.
* Persistence logic and core business logic are mixed together in your DNA, because at its heart, Holochain is a framework for data validation. However, you usually don’t need a lot of your business logic in your DNA—just enough to encode the ‘rules of the game’ for your application.
* As with microservices, Holochain lends itself to combining small, reusable components into large applications.

## Learn more

> If you want to go from theory to practice, you can learn hands-on how applications talk to hApps by reading and following the [hApp client call tutorial](https://github.com/holochain/happ-client-call-tutorial), which has examples in Typescript and Rust.

Reviewer: You can see in my git commit my rationale... 
I looked in to those links and they are very lofty and off topic. I think it'd be good to take this chance to ground the technical reader.

Author: Sounds totally reasonable to me. Thanks for checking those articles.
